File: native/MorveroFeedback.swift
// MorveroFeedback.swift — drop-in Morvero client for iOS/macOS. v1.0.0
//
// Single file, no dependencies, iOS 13+ / macOS 10.15+. Add this file to your
// app target, then:
//
//   MorveroFeedback.configure(origin: "https://morvero.com",
//                             key: "fbk_your_product_key",
//                             appId: "com.company.app")        // must match the key's allowed app IDs (if set)
//   MorveroFeedback.trackScreen("/invoices")                    // on screen appear
//   MorveroFeedback.trackAction("export-csv")                   // on feature use
//   MorveroFeedback.submitFeedback(rating: 4, comment: "…")     // from your feedback UI
//
// Privacy: the visitor id is a random, meaningless string stored only in
// UserDefaults — no device identifiers, no personal data. The optional cohort
// is a coarse group label (e.g. "finance"); never pass usernames or ids.
// Every call is fire-and-forget on a background queue and will never throw,
// block, or crash the host app.

import Foundation

public enum MorveroFeedback {
    private static var origin = ""
    private static var key = ""
    private static var appId = ""
    private static var cohort = ""
    private static let queue = DispatchQueue(label: "com.morvero.feedback", qos: .utility)

    private static var visitorId: String {
        let d = UserDefaults.standard
        if let v = d.string(forKey: "morvero_visitor") { return v }
        let v = "v_" + UUID().uuidString.lowercased().replacingOccurrences(of: "-", with: "")
        d.set(v, forKey: "morvero_visitor")
        return v
    }

    public static func configure(origin: String, key: String, appId: String = "", cohort: String = "") {
        self.origin = origin.hasSuffix("/") ? String(origin.dropLast()) : origin
        self.key = key
        self.appId = appId
        self.cohort = cohort
    }

    /// Record a screen view. Path is a stable route-like string ("/invoices").
    public static func trackScreen(_ path: String, title: String = "") {
        post("/api/collect/pageview", body: [
            "key": key, "path": path, "title": title, "visitorId": visitorId, "cohort": cohort,
        ])
    }

    /// Record a feature/function use; cost (USD) is for AI-agent telemetry.
    public static func trackAction(_ name: String, path: String = "/", cost: Double = 0) {
        var body: [String: Any] = ["key": key, "action": name, "path": path, "visitorId": visitorId]
        if cost > 0 { body["cost"] = cost }
        post("/api/collect/action", body: body)
    }

    /// Submit a 1–5 rating with an optional comment. Fetches the short-lived
    /// submission token for you and retries once on an expired token.
    public static func submitFeedback(rating: Int, comment: String = "", path: String = "/",
                                      completion: ((Bool) -> Void)? = nil) {
        queue.async {
            func attempt(_ retriesLeft: Int) {
                fetchToken { token in
                    guard let token = token else { completion?(false); return }
                    request("/api/collect/feedback", body: [
                        "key": key, "token": token, "rating": rating, "comment": comment,
                        "path": path, "visitorId": visitorId, "platform": platformName(),
                        "trigger": "manual", "cohort": cohort,
                    ]) { status in
                        if status == 401 && retriesLeft > 0 { attempt(retriesLeft - 1) }
                        else { completion?(status == 201) }
                    }
                }
            }
            attempt(1)
        }
    }

    // MARK: - internals

    private static func platformName() -> String {
        #if os(iOS)
        return "ios"
        #else
        return "api"
        #endif
    }

    private static func fetchToken(_ done: @escaping (String?) -> Void) {
        guard var comps = URLComponents(string: origin + "/api/widget-config") else { done(nil); return }
        comps.queryItems = [URLQueryItem(name: "key", value: key)]
        guard let url = comps.url else { done(nil); return }
        var req = URLRequest(url: url)
        if !appId.isEmpty { req.setValue(appId, forHTTPHeaderField: "X-Morvero-App-Id") }
        URLSession.shared.dataTask(with: req) { data, _, _ in
            let token = data.flatMap { try? JSONSerialization.jsonObject(with: $0) as? [String: Any] }?["token"] as? String
            done(token)
        }.resume()
    }

    private static func post(_ path: String, body: [String: Any]) {
        queue.async { request(path, body: body, done: { _ in }) }
    }

    private static func request(_ path: String, body: [String: Any], done: @escaping (Int) -> Void) {
        guard !origin.isEmpty, !key.isEmpty, let url = URL(string: origin + path) else { done(0); return }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !appId.isEmpty { req.setValue(appId, forHTTPHeaderField: "X-Morvero-App-Id") }
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)
        URLSession.shared.dataTask(with: req) { _, resp, _ in
            done((resp as? HTTPURLResponse)?.statusCode ?? 0)
        }.resume()
    }
}

File: native/MorveroFeedback.kt
// MorveroFeedback.kt — drop-in Morvero client for Android. v1.0.0
//
// Single file, no dependencies beyond the Android SDK (HttpURLConnection +
// org.json). minSdk 21. Add this file to your app module, then:
//
//   MorveroFeedback.configure(context, origin = "https://morvero.com",
//                             key = "fbk_your_product_key",
//                             appId = "com.company.app")   // must match the key's allowed app IDs (if set)
//   MorveroFeedback.trackScreen("/invoices")                // on screen shown
//   MorveroFeedback.trackAction("export-csv")               // on feature use
//   MorveroFeedback.submitFeedback(4, "Love it")            // from your feedback UI
//
// Privacy: the visitor id is a random, meaningless string stored only in the
// app's SharedPreferences — no device identifiers, no personal data. The
// optional cohort is a coarse group label (e.g. "finance"); never pass
// usernames or ids. Every call runs fire-and-forget on a single background
// thread and will never throw, block, or crash the host app.

package com.morvero.feedback

import android.content.Context
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID
import java.util.concurrent.Executors

object MorveroFeedback {
    private var origin = ""
    private var key = ""
    private var appId = ""
    private var cohort = ""
    private var visitorId = ""
    private val executor = Executors.newSingleThreadExecutor { r -> Thread(r, "morvero-feedback").apply { isDaemon = true } }

    @JvmStatic
    @JvmOverloads
    fun configure(context: Context, origin: String, key: String, appId: String = "", cohort: String = "") {
        this.origin = origin.trimEnd('/')
        this.key = key
        this.appId = appId
        this.cohort = cohort
        val prefs = context.applicationContext.getSharedPreferences("morvero", Context.MODE_PRIVATE)
        visitorId = prefs.getString("visitor", null) ?: ("v_" + UUID.randomUUID().toString().replace("-", "")).also {
            prefs.edit().putString("visitor", it).apply()
        }
    }

    /** Record a screen view. Path is a stable route-like string ("/invoices"). */
    @JvmStatic
    @JvmOverloads
    fun trackScreen(path: String, title: String = "") = post("/api/collect/pageview", JSONObject()
        .put("key", key).put("path", path).put("title", title)
        .put("visitorId", visitorId).put("cohort", cohort))

    /** Record a feature/function use; cost (USD) is for AI-agent telemetry. */
    @JvmStatic
    @JvmOverloads
    fun trackAction(name: String, path: String = "/", cost: Double = 0.0) = post("/api/collect/action", JSONObject()
        .put("key", key).put("action", name).put("path", path).put("visitorId", visitorId)
        .apply { if (cost > 0) put("cost", cost) })

    /**
     * Submit a 1–5 rating with an optional comment. Fetches the short-lived
     * submission token for you and retries once on an expired token.
     */
    @JvmStatic
    @JvmOverloads
    fun submitFeedback(rating: Int, comment: String = "", path: String = "/", callback: ((Boolean) -> Unit)? = null) {
        executor.execute {
            var ok = false
            try {
                var attempts = 0
                while (attempts < 2) {
                    attempts++
                    val token = fetchToken() ?: break
                    val status = request("/api/collect/feedback", JSONObject()
                        .put("key", key).put("token", token).put("rating", rating).put("comment", comment)
                        .put("path", path).put("visitorId", visitorId).put("platform", "android")
                        .put("trigger", "manual").put("cohort", cohort))
                    if (status == 401) continue
                    ok = status == 201
                    break
                }
            } catch (_: Exception) { /* never break the host app */ }
            callback?.invoke(ok)
        }
    }

    // --- internals ---

    private fun fetchToken(): String? = try {
        val conn = URL("$origin/api/widget-config?key=$key").openConnection() as HttpURLConnection
        if (appId.isNotEmpty()) conn.setRequestProperty("X-Morvero-App-Id", appId)
        conn.connectTimeout = 8000; conn.readTimeout = 8000
        val body = conn.inputStream.bufferedReader().readText()
        conn.disconnect()
        JSONObject(body).optString("token").ifEmpty { null }
    } catch (_: Exception) { null }

    private fun post(path: String, body: JSONObject) {
        if (origin.isEmpty() || key.isEmpty()) return
        executor.execute { try { request(path, body) } catch (_: Exception) { /* fire and forget */ } }
    }

    private fun request(path: String, body: JSONObject): Int {
        val conn = URL(origin + path).openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.doOutput = true
        conn.setRequestProperty("Content-Type", "application/json")
        if (appId.isNotEmpty()) conn.setRequestProperty("X-Morvero-App-Id", appId)
        conn.connectTimeout = 8000; conn.readTimeout = 8000
        conn.outputStream.use { it.write(body.toString().toByteArray()) }
        val status = conn.responseCode
        conn.disconnect()
        return status
    }
}
